language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/ExtractToFuncStatementAnalyzer.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2022, WSO2 LLC. (http://wso2.com) All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.codeaction;
+
+import io.ballerina.compiler.api.SemanticModel;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.syntax.tree.AssignmentStatementNode;
+import io.ballerina.compiler.syntax.tree.CompoundAssignmentStatementNode;
+import io.ballerina.compiler.syntax.tree.IfElseStatementNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.ReturnStatementNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.Token;
+import io.ballerina.compiler.syntax.tree.VariableDeclarationNode;
+import org.ballerinalang.langserver.common.utils.PositionUtil;
+import org.eclipse.lsp4j.Range;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Node visitor for extract to function code action. This will determine,
+ * 1. Symbols which get assigned to some expression inside the selected range
+ * 2. Symbols which are declared inside the selected range
+ * 3. Whether the selected range is extractable
+ * 4. Nodes which are inside the selected range which should be extracted
+ *
+ * @since 2201.2.1
+ */
+public class ExtractToFuncStatementAnalyzer extends NodeVisitor {
```

Comment:
This visitor evaluates multiple attributes and characteristics (not only extractability but also updated symbols etc) hence did not use `xxxxStatementFinder` which implies to finding statements. Also this is only unique to the extract to function (not even other extract to xxx code action), hence IMO this name suits better. Didn't use the full code action name `Extract to function` since it's making the name longer. But maybe we can use the longer version